app/models/game.rb

```diff
@@ -16,24 +16,44 @@ def whos_turn?
     return black_player_id if turn_number.odd?
   end
 
-  # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
   def determine_check
-    if turn_number.even?
-      opponent_pieces = pieces.where(user_id: black_player_id)
-      king = pieces.find_by(user_id: white_player_id, type: "King")
-      opponent_pieces.each do |piece|
-        return true if piece.valid_move?(king.row_position, king.col_position)
-      end
-    else
-      opponent_pieces = pieces.where(user_id: white_player_id)
-      king = pieces.find_by(user_id: black_player_id, type: "King")
-      opponent_pieces.each do |piece|
-        return true if piece.valid_move?(king.row_position, king.col_position)
-      end
+    return check(white_player_id, black_player_id) if turn_number.even?
+    check(black_player_id, white_player_id)
+  end
+
+  def check(id, opponent_id)
+    opponent_pieces = pieces.where(user_id: opponent_id)
+    king = pieces.find_by(user_id: id, type: "King")
+    opponent_pieces.each do |piece|
+      return true if piece.valid_move?(king.row_position, king.col_position)
     end
     false
   end
 
+  def determine_checkmate
+    return checkmate(white_player_id) if turn_number.even? && determine_check
+    checkmate(black_player_id) if turn_number.odd? && determine_check
+  end
+
+  # rubocop:disable Metrics/MethodLength
+  def checkmate(id)
+    current_pieces = pieces.where(user_id: id)
+    check_status = true
+    current_pieces.each do |piece|
+      8.times do |row|
+        8.times do |col|
+          next if piece.valid_move?(row, col)
+          Piece.transaction do
+            piece.move_to!(row, col)
+            check_status = false if determine_check == false
+            fail ActiveRecord::Rollback
+          end
```

Comment:
@jonathanpike: thanks for finishing it up! 
I have one confusion about line 45. If the piece can move to the (row, col) we skip lines 46-50 (we don't try to move it there and call determine_check)? 
And if it can't move there (valid_move returns false), we do try to move it there and call determine_check?
Shouldn't it be the other way around? -  next if !piece.valid_move?(row, col)
I probably miss something :)